pkg/planner/core/plan_cache_utils.go

```diff
@@ -319,6 +319,54 @@ func NewPlanCacheKey(sctx sessionctx.Context, stmt *PlanCacheStmt) (key, binding
 	hash = append(hash, hack.Slice(strconv.FormatBool(variable.VarTiDBSuperReadOnly.Load()))...)
 	// expr-pushdown-blacklist can affect query optimization, so we need to consider it in plan cache.
 	hash = codec.EncodeInt(hash, expression.ExprPushDownBlackListReloadTimeStamp.Load())
+
+	// whether this query has sub-query
+	if stmt.QueryFeatures.hasSubquery {
+		hash = append(hash, '1')
+	} else {
+		hash = append(hash, '0')
+	}
+
+	// this variable might affect the plan
+	if vars.ForeignKeyChecks {
+		hash = append(hash, '1')
+	} else {
+		hash = append(hash, '0')
+	}
+
+	// "limit ?" can affect the cached plan: "limit 1" and "limit 10000" should use different plans.
+	if len(stmt.QueryFeatures.limits) > 0 {
+		if !vars.EnablePlanCacheForParamLimit {
+			return "", "", false, "plan cache is disabled for param limit", nil
+		}
+		hash = append(hash, '|')
+		for _, node := range stmt.QueryFeatures.limits {
+			if node.Count != nil {
+				if count, isParamMarker := node.Count.(*driver.ParamMarkerExpr); isParamMarker {
+					typeExpected, val := CheckParamTypeInt64orUint64(count)
+					if !typeExpected {
+						return "", "", false, "unexpected value after LIMIT", err
+					}
+					if val > MaxCacheableLimitCount {
+						return "", "", false, "limit count is too large", err
+					}
+					hash = codec.EncodeUint(hash, val)
+				}
+			}
+			if node.Offset != nil {
+				if offset, isParamMarker := node.Offset.(*driver.ParamMarkerExpr); isParamMarker {
+					typeExpected, val := CheckParamTypeInt64orUint64(offset)
+					if !typeExpected {
+						return "", "", false, "unexpected value after LIMIT", err
+					}
+					hash = codec.EncodeUint(hash, val)
+				}
+			}
```

Comment:
This code contains some redundant parts and appears to be simplifiable.

Reply:
OK, I've simplified it.